lib/core/socket-handler.ts

```diff
@@ -68,60 +68,72 @@ export class SocketHandler {
             let badMessage = false;
             if (messageIDIndex < 0) {
                 badMessage = true;
             }
 
-            let message = this.buffer.substring(0, messageIDIndex);
+            const message = this.buffer.slice(0, messageIDIndex);
             // Grab the message ID - it precedes the delimiter
-            let messageIDString = this.buffer.substring(delimiterIndex - Global.MessageIDLength, delimiterIndex);
-            let messageID: number = parseInt(messageIDString);
+            const messageIDString = dataString.slice(delimiterIndex - Global.MessageIDLength, delimiterIndex);
+            const messageID: number = parseInt(messageIDString);
             if (isNaN(messageID) || (messageID + "").length < 13) {
                 badMessage = true;
             }
 
             if (badMessage) {
                 LoggingHelper.error(Logger, "Bad message received: " + dataString);
             } else {
-                LoggingHelper.debug(Logger, "DATA READ " + this.remoteEndPoint() + " ID: " + messageID +  " MSG: " + StringUtil.prettyPrint(message));
+                if (typeof message === "string") {
+                    LoggingHelper.debug(Logger, "DATA READ " + this.remoteEndPoint() + " ID: " + messageID +  " MSG: " + StringUtil.prettyPrint(message));
+                } else {
+                    LoggingHelper.debug(Logger, "DATA READ " + this.remoteEndPoint() + " ID: " + messageID +  " MSG: < Binary Data >");
+                }
                 this.onMessage(message, messageID);
             }
 
             this.buffer = this.buffer.slice(delimiterIndex + Global.MessageDelimiter.length);
 
             // If we have received more than one packet at a time, handle it recursively
-            if (this.buffer.indexOf(Global.MessageDelimiter) !== -1) {
+            if (this.buffer.toString().indexOf(Global.MessageDelimiter) !== -1) {
                 this.handleData(null);
             }
         }
     }
 
-    public send(message: string, messageID?: number) {
+    public send(message: string | Buffer, messageID?: number) {
         // If the socket was already closed, do not write anything
         if (this.socket === null) {
             LoggingHelper.warn(Logger, "Writing message to closed socket: " + messageID);
             return;
         }
 
-        LoggingHelper.debug(Logger, "DATA SENT " + this.remoteEndPoint() + " SEQUENCE: " + messageID + " " + StringUtil.prettyPrint(message));
+        LoggingHelper.debug(Logger, "DATA SENT " + this.remoteEndPoint() + " SEQUENCE: " + messageID + " " + StringUtil.prettyPrint(message.toString()));
 
         // If no message ID is specified, just grab a timestamp
         if (messageID === undefined || messageID === null) {
             messageID = new Date().getTime();
         }
         // Use TOKEN as message delimiter
-        message = message + messageID + Global.MessageDelimiter;
+        if (typeof message === "string") {
```

Comment:
Adding the delimiter correctly can also be encapsulated in a SocketMessage class